File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.js
const fs = require('fs');
let [N, nums, operators] = fs
  .readFileSync(0, 'utf8')
  .trim()
  .split('\n');

N = +N;
nums = nums.split(' ').map((v) => +v);
operators = operators.split(' ').map((v) => +v);

function solution(N, nums, operators) {
  let max = -Infinity;
  let min = Infinity;

  function calc(perm) {
    const value = nums.reduce((acc, cur, idx) => {
      if (idx > 0) {
        switch (perm[idx - 1]) {
          case 0:
            return acc + cur;
          case 1:
            return acc - cur;
          case 2:
            return acc * cur;
          case 3:
            return ~~(acc / cur);
        }
      } else {
        return acc + cur;
      }
    });

    if (value < min) min = value;
    if (value > max) max = value;
  }

  function backtracking(perm, operators) {
    if (perm.length === N - 1) {
      calc(perm);
      return;
    }

    operators.forEach((operator, idx) => {
      if (operator > 0) {
        let new_operators = [...operators];
        new_operators[idx] -= 1;
        backtracking([...perm, idx], new_operators);
      }
    });
  }

  backtracking([], operators);

  return [max, min].join('\n');
}

console.log(solution(N, nums, operators));

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
N = int(input())
nums = list(map(int, input().split()))
opperators = list(map(int, input().split()))


def solution(N, nums, opperators):
    min = 1e9 + 1
    max = -1e9 - 1

    def calc(perm):
        nonlocal min
        nonlocal max
        value = 0

        for idx, num in enumerate(nums):
            if idx == 0:
                value += num
            else:
                if perm[idx - 1] == 0:

                    value += num
                elif perm[idx - 1] == 1:
                    value -= num
                elif perm[idx - 1] == 2:
                    value *= num
                else:
                    value = int(value / num)

        if value < min:
            min = value
        if value > max:
            max = value

    def backtracking(perm, opperators):
        if len(perm) == N - 1:
            calc(perm)
            return

        for idx, opperator in enumerate(opperators):
            if opperator > 0:
                new_opperators = opperators[:]
                new_opperators[idx] -= 1
                new_perm = perm[:]
                new_perm.append(idx)
                backtracking(new_perm, new_opperators)

    backtracking([], opperators)

    return f"{max}\n{min}"


print(solution(N, nums, opperators))
